miri-script/src/coverage.rs

```diff
@@ -0,0 +1,84 @@
+use std::ffi::OsString;
+
+use anyhow::{Context, Result};
+use xshell::cmd;
+
+use crate::util::MiriEnv;
+
+/// CoverageReport can generate code coverage reports for miri.
+pub struct CoverageReport {
+
+    /// path is a temporary directory where coverage artifacts will be stored.
+    path: tempfile::TempDir,
+}
+
+impl CoverageReport {
+
+    /// new creates a new CoverageReport
+    /// 
+    /// # Errors
+    /// 
+    /// An error will be returned if a temporary directory could not be created.
+    pub fn new() -> Result<Self> {
+        Ok(Self { path: tempfile::TempDir::new()? })
+    }
+
+    /// add_env_vars will add the required environment variables to MiriEnv `e`.
+    pub fn add_env_vars(&self, e: &mut MiriEnv) -> Result<()> {
+        let rustflags = e.sh.var("RUSTFLAGS")?;
+        let rustflags = format!("{rustflags} -C instrument-coverage");
+        e.sh.set_var("RUSTFLAGS", rustflags);
+
+        let file_template = self.path.path().join("default_%m_%p.profraw");
+        e.sh.set_var("LLVM_PROFILE_FILE", file_template);
+        Ok(())
+    }
+
+    /// show_coverage_report will print coverage information using the artifact
+    /// files in `self.path`. 
+    pub fn show_coverage_report(&self, e: &MiriEnv) -> Result<()> {
+        let profraw_files: Vec<_> = self.profraw_files()?;
+
+        let mut profdata_bin = e.libdir.clone();
+        profdata_bin.push("..");
+        profdata_bin.push("bin");
+        profdata_bin.push("llvm-profdata");
+
+        let merged_file = self.path.path().join("merged.profdata");
+
+        // Merge the profraw files
+        let profraw_files_cloned = profraw_files.iter();
+        cmd!(e.sh, "{profdata_bin} merge -sparse {profraw_files_cloned...} -o {merged_file}")
+            .quiet()
+            .run()?;
+
+        // Create the coverage report.
+        let mut cov_bin = e.libdir.clone();
+        cov_bin.push("..");
+        cov_bin.push("bin");
+        cov_bin.push("llvm-cov");
```

Comment:
Can this be written nicer with the `path!` macro?

Reply:
yes, thanks!